34.md

```diff
@@ -0,0 +1,84 @@
+NIP-34
+======
+
+`git` stuff
+-----------
+
+`draft` `optional`
+
+This NIP defines all the ways code collaboration using and adjacent to [`git`](https://git-scm.com/) can be done using Nostr.
+
+## Repository announcements
+
+Git repositories are hosted in Git-enabled servers, but their existence can be announced using Nostr events, as well as their willingness to receive patches, bug reports and comments in general.
+
+```jsonc
+{
+  "kind": 30617,
+  "content": "",
+  "tags": [
+    ["d", "<repo-id>"],
+    ["name", "<human-readable project name>"],
+    ["description", "brief human-readable project description>"],
+    ["head", "<sha1-hex>"], // the current repository commit head, as hex
+    ["web", "<url for browsing>"], // a webpage url, if the git server being used provides such a thing
+    ["clone", "<url for git-cloning>"], // a url to be given to `git clone` so anyone can clone it
+    ["patches", "<relay-url>"], // this signals that patches are welcome and will be searched for in this relay
+    ["issues", "<relay-url>"] // this signals that bug reports, questions, comments and so on are welcome
+  ]
+}
+```
+
+The tags `web`, `clone`, `patches` and `issues` can be specified multiple times. Except `d`, all tags are optional.
+
+## Patches
+
+Patches can be sent by anyone to any repository. Patches to a specific repository SHOULD be sent to the relays specified in that repository's announcement event's `"patches"` tag. Patch events SHOULD include an `a` tag pointing to that repository's announcement address.
+
+```jsonc
+{
+  "kind": 1617,
+  "content": "<patch>", // contents of <git format-patch>
+  "tags": [
+    ["a", "30617:<base-repo-owner-pubkey>:<base-repo-id>"],
+    ["p", "<repository-owner>"]
+  ]
```

Comment:
Probably also want the commit that this patch is based on?

Reply:
Good point. How does `git send-email` handle that?

OK, I just checked and it doesn't add that automatically. How do people do it in practice?

Reply:
> How do people do it in practice?

For projects that prefer patches rather than commits, the maintainers usually just try to apply the patch on the latest master and if it conflicts they reply and tell you to rebase it.

But for the use case here, that would just result in no one sending any more patches, so the best way to handle it is just copy/paste the latest HEAD that you're working on top of, and add it to a `base` tag (that's what I was using for nostrocket anyway).

Maintainers can then manually apply the patch on top of the correct commit and resolve any conflicts manually.

All of this just makes me think we should be using native git most of the time though rather than patches.

Reply:
> OK, I just checked and it doesn't add that automatically. How do people do it in practice?

you use `git-format-patch --base=HEAD^ HEAD`

this includes base-commit info in the patch itself, so you can know where the commit's parent.

it just includes it as text at the bottom of the patch: 

`base-commit: c67741983e3f07f2386eaa388cb8a1475e8e047`

Reply:
it might be interesting to make this explicit in a tag though!

Reply:
I think for practical purposes this is better left to be read manually by the maintainer.

But in any case I've added this as an optional tag that can be used when people want a stable commit id, so it can be used for other purposes too I guess.